### PR TITLE
PW-4495 Added support for client certificate authentication

### PR DESCRIPTION
**Pull request by saquibsayyad**
**Description**
Added support for client certificate authentication. (Right now only supported in live)

**Tested scenarios**
Tested in LIVE

### CODE REVIEW

File: src/main/java/com/adyen/httpclient/AdyenHttpClient.java [hunk 1]
```diff
@@ -181,35 +183,63 @@ private URI createUri(String endpoint, Map<String, String> params) throws HTTPCl
...
-
-                // Install the terminal certificate trust manager
-                SSLContext sc = SSLContext.getInstance(SSL);
-                sc.init(null, trustManagers, new java.security.SecureRandom());
-
-                SSLConnectionSocketFactory sslConSocFactory = new SSLConnectionSocketFactory(sc, createHostnameVerifier(config.getEnvironment()));
-                httpClientBuilder.setSSLSocketFactory(sslConSocFactory);
-            } catch (GeneralSecurityException | IOException e) {
-                throw new HTTPClientException("Error loading certificate from path", e);
+            KeyStore keyStore = KeyStore.getInstance(JAVA_KEYSTORE);
```
Comment (Aleffio) [inline src/main/java/com/adyen/httpclient/AdyenHttpClient.java:203]: should we use `KeyStore.getDefaultType` like suggested in #567?
```suggestion
            KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
```

Comment (saquibsayyad) [reply src/main/java/com/adyen/httpclient/AdyenHttpClient.java]: > should we use `KeyStore.getDefaultType` like suggested in #567?

corrected it, thanks

File: src/main/java/com/adyen/httpclient/AdyenHttpClient.java [hunk 2]
```diff
@@ -69,7 +69,9 @@
 
     private static final String CHARSET = "UTF-8";
     private static final String TERMINAL_CERTIFICATE_ALIAS = "TerminalCertificate";
+    private static final String JAVA_KEYSTORE = "JKS";
```
Comment (Aleffio) [inline src/main/java/com/adyen/httpclient/AdyenHttpClient.java:72]: I forgot this one :)
```suggestion
```

**Review by Aleffio — APPROVED** [review]

File: src/main/java/com/adyen/httpclient/AdyenHttpClient.java (hunk 2 above)
Comment (saquibsayyad) [reply src/main/java/com/adyen/httpclient/AdyenHttpClient.java]: > I forgot this one :)

my bad, i just applied your suggestion using Github UI instead of doing it in code.

**Review by Aleffio — DISMISSED** [review]

File: src/main/java/com/adyen/Client.java [hunk 3]
```diff
@@ -77,6 +81,33 @@ public Client(String username, String password, Environment environment, String
...
+     * Note: Client certificate authentication is only applicable for PAL and Checkout services in LIVE,
+     * Other services will just use API key for authentication.
+     * @param trustStore Trust store containing server certificate
+     * @param clientKeyStore Client Key store containing client certificate and key
+     * @param clientKeyStorePassword Password for client key store
+     * @param apiKey Adyen API Key
+     * @param region Data center region (EU/US/AU), default EU if not provided
+     */
+    public Client(KeyStore trustStore, KeyStore clientKeyStore, String clientKeyStorePassword, String apiKey, Region region) {
+        this();
```
Comment (rkewlani) [inline src/main/java/com/adyen/Client.java:95]: minor: this(String apiKey, Environment environment) can be used here.

File: src/test/java/com/adyen/util/CertificateUtilTest.java [hunk 4]
```diff
@@ -50,4 +55,19 @@ public void testLoadCertificateFromInputStream() throws CertificateException {
 
         Assert.assertNotNull(certificate);
     }
+
```
Comment (rkewlani) [inline src/test/java/com/adyen/util/CertificateUtilTest.java:58]: minor: more tests can be added for other negative scenarios like if resource.getPath() is null or keystore type is null etc.

**Review by rkewlani — APPROVED** [review]

File: src/main/java/com/adyen/Client.java (hunk 3 above)
Comment (saquibsayyad) [reply src/main/java/com/adyen/Client.java:95]: thanks, will update it,

File: src/test/java/com/adyen/util/CertificateUtilTest.java (hunk 4 above)
Comment (saquibsayyad) [reply src/test/java/com/adyen/util/CertificateUtilTest.java:58]: Added null check and more tests.

File: src/main/java/com/adyen/Client.java (hunk 3 above)
Comment (Aleffio) [reply src/main/java/com/adyen/Client.java]: `this(String apiKey, Environment environment)` sets also the other endpoints(marketpay, hpp,...) that do not support client certificate auth.
Do we really want the merchants to use the same client with the certificate, also for "non certificate" endpoints?

Comment (saquibsayyad) [reply src/main/java/com/adyen/Client.java]: `this(apiKey, Environment.LIVE)` was just to replace individual `this(),         this.setEnvironment(Environment.LIVE, null); and this.config.setApiKey(apiKey);` so the behaviour is same.

However if we don't want to setup other endpoints then i can remove the `this.setEnvironment(Environment.LIVE, null)` and just set PAL and Checkout endpoints.

**Review by Aleffio — APPROVED** [review]

**Review by rkewlani — APPROVED** [review]
